SU2_CFD/src/solvers/CFEASolver.cpp

```diff
@@ -0,0 +1,3774 @@
+
+  const bool topology_mode = config->GetTopology_Optimization();
+  const su2double simp_exponent = config->GetSIMP_Exponent();
+  const su2double simp_minstiff = config->GetSIMP_MinStiffness();
+
+  /*--- Start OpenMP parallel region. ---*/
+
+  SU2_OMP_PARALLEL
+  {
+#ifdef HAVE_OMP
+    /*--- Chunk size is at least OMP_MIN_SIZE and a multiple of the color group size. ---*/
+    auto chunkSize = roundUpDiv(OMP_MIN_SIZE, ColorGroupSize)*ColorGroupSize;
+
+    /*--- Loop over element colors. ---*/
+    for (auto color : ElemColoring)
+    {
+      SU2_OMP_FOR_DYN(chunkSize)
+      for(auto k = 0ul; k < color.size; ++k) {
+
+        auto iElem = color.indices[k];
+#else
+    /*--- Natural coloring. ---*/
+    {
+      for (auto iElem = 0ul; iElem < nElement; iElem++) {
+#endif
+        unsigned short iNode, jNode, iDim, iVar;
+
+        int thread = omp_get_thread_num();
+
+        /*--- Convert VTK type to index in the element container. ---*/
+        int EL_KIND;
+        unsigned short nNodes;
+        GetElemKindAndNumNodes(geometry->elem[iElem]->GetVTK_Type(), EL_KIND, nNodes);
+
+        /*--- Each thread needs a dedicated element. ---*/
+        CElement* element = element_container[FEA_TERM][EL_KIND+thread*MAX_FE_KINDS];
+
+        /*--- For the number of nodes, get the coordinates and cache the point indices. ---*/
+        unsigned long indexNode[MAXNNODE];
+
+        for (iNode = 0; iNode < nNodes; iNode++) {
+
+          indexNode[iNode] = geometry->elem[iElem]->GetNode(iNode);
+
+          for (iDim = 0; iDim < nDim; iDim++) {
+            su2double val_Coord = Get_ValCoord(geometry, indexNode[iNode], iDim);
+            su2double val_Sol = nodes->GetSolution(indexNode[iNode],iDim) + val_Coord;
+            element->SetRef_Coord(iNode, iDim, val_Coord);
+            element->SetCurr_Coord(iNode, iDim, val_Sol);
+          }
+        }
+
+        /*--- In topology mode determine the penalty to apply to the stiffness. ---*/
+        su2double simp_penalty = 1.0;
+        if (topology_mode) {
+          su2double density = element_properties[iElem]->GetPhysicalDensity();
+          simp_penalty = simp_minstiff+(1.0-simp_minstiff)*pow(density,simp_exponent);
+        }
+
+        /*--- Set the properties of the element ---*/
+        element->Set_ElProperties(element_properties[iElem]);
+
+        /*--- Compute the components of the jacobian and the stress term, one numerics per thread. ---*/
+        int NUM_TERM = thread*MAX_TERMS + element_properties[iElem]->GetMat_Mod();
+
+        numerics[NUM_TERM]->Compute_Tangent_Matrix(element, config);
+
+        /*--- Update residual and stiffness matrix with contributions from the element. ---*/
+        for (iNode = 0; iNode < nNodes; iNode++) {
+
+          auto Ta = element->Get_Kt_a(iNode);
+          for (iVar = 0; iVar < nVar; iVar++)
+            LinSysRes[indexNode[iNode]*nVar+iVar] -= simp_penalty*Ta[iVar];
```

Comment:
In a number of places we now write directly to CSysVector / CSysMatrix instead of populating a temporary first.

Reply:
I like it, it's cleaner like this.